File: dodam-application/dodam-rest-api/src/main/java/b1nd/dodam/restapi/wakeupsong/presentation/WakeupSongController.java
package b1nd.dodam.restapi.wakeupsong.presentation;

import b1nd.dodam.melon.chart.client.data.res.ChartRes;
import b1nd.dodam.restapi.support.data.Response;
import b1nd.dodam.restapi.support.data.ResponseData;
import b1nd.dodam.restapi.wakeupsong.application.WakeupSongUseCase;
import b1nd.dodam.restapi.wakeupsong.application.data.req.ApplyWakeupSongBySearchReq;
import b1nd.dodam.restapi.wakeupsong.application.data.req.ApplyWakeupSongReq;
import b1nd.dodam.restapi.wakeupsong.application.data.res.WakeupSongRes;
import b1nd.dodam.youtube.video.client.data.YoutubeRes;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.concurrent.CompletableFuture;

@RestController
@RequestMapping("/wakeup-song")
@RequiredArgsConstructor
@Slf4j
public class WakeupSongController {
    private final WakeupSongUseCase wakeupSongUseCase;

    @GetMapping("/allowed")
    public ResponseData<List<WakeupSongRes>> getAllowedWakeupSongByPlayDate(
            @RequestParam int year,
            @RequestParam int month,
            @RequestParam int day
    ) {
        return wakeupSongUseCase.getAllowedWakeupSong(year, month, day);
    }

    @GetMapping("/pending")
    public ResponseData<List<WakeupSongRes>> getPendingWakeupSong() {
        return wakeupSongUseCase.getPendingWakeupSong();
    }

    @GetMapping("/my")
    public ResponseData<List<WakeupSongRes>> getMyWakeupSong() {
        return wakeupSongUseCase.getMyWakeupSong();
    }

    @PostMapping
    public CompletableFuture<Response> createWakeupSong(@RequestBody @Valid ApplyWakeupSongReq req) {
        return wakeupSongUseCase.createWakeupSong(req.videoUrl());
    }

    @PostMapping("/keyword")
    public CompletableFuture<Response> createWakeupSongByYoutubeSearch(@RequestBody @Valid ApplyWakeupSongBySearchReq req) {
        return wakeupSongUseCase.createWakeupSongByYoutubeSearch(req);
    }

    @GetMapping("/search")
    public ResponseData<List<YoutubeRes>> getYoutubeVideo(@RequestParam String keyword) {
        return wakeupSongUseCase.getYoutubeList(keyword);
    }

    @PatchMapping("/allow/{id}")
    public Response allowWakeupSong(@PathVariable int id) {
        return wakeupSongUseCase.allow(id);
    }

    @PatchMapping("/deny/{id}")
    public Response denyWakeupSong(@PathVariable int id) {
        return wakeupSongUseCase.deny(id);
    }

    @DeleteMapping("/{id}")
    public Response deleteWakeupSong(@PathVariable int id) {
        return wakeupSongUseCase.delete(id);
    }

    @DeleteMapping("/my/{id}")
    public Response deleteMyWakeupSong(@PathVariable int id) {
        return wakeupSongUseCase.deleteMyWakeupSong(id);
    }

    @GetMapping("/chart")
    public CompletableFuture<ResponseData<List<ChartRes>>> getChart() {
        return wakeupSongUseCase.getChartList();
    }

}

File: services/service-wakeup-song/src/main/kotlin/com/b1nd/dodamdodam/wakeupsong/presentation/WakeupSongController.kt
package com.b1nd.dodamdodam.wakeupsong.presentation

import com.b1nd.dodamdodam.core.security.annotation.authentication.UserAccess
import com.b1nd.dodamdodam.core.security.passport.enumerations.RoleType
import com.b1nd.dodamdodam.wakeupsong.application.WakeupSongUseCase
import com.b1nd.dodamdodam.wakeupsong.application.data.request.ApplyWakeupSongByKeywordRequest
import com.b1nd.dodamdodam.wakeupsong.application.data.request.ApplyWakeupSongRequest
import org.springframework.web.bind.annotation.*

@RestController
class WakeupSongController(
    private val wakeupSongUseCase: WakeupSongUseCase
) {

    @UserAccess(hasAnyRoleOnly = true)
    @GetMapping("/my")
    fun getMyWakeupSongs() = wakeupSongUseCase.getMyWakeupSongs()

    @UserAccess(roles = [RoleType.STUDENT])
    @GetMapping("/allowed")
    fun getAllowed(
        @RequestParam year: Int,
        @RequestParam month: Int,
        @RequestParam day: Int
    ) = wakeupSongUseCase.getAllowed(year, month, day)

    @UserAccess(roles = [RoleType.TEACHER])
    @GetMapping("/pending")
    fun getPending() = wakeupSongUseCase.getPending()

    @UserAccess(hasAnyRoleOnly = true)
    @GetMapping("/chart")
    fun getMelonChart() = wakeupSongUseCase.getMelonChart()

    @UserAccess(hasAnyRoleOnly = true)
    @GetMapping("/search")
    fun searchYouTube(@RequestParam keyword: String) = wakeupSongUseCase.searchYouTube(keyword)

    @UserAccess(roles = [RoleType.STUDENT])
    @PostMapping
    fun applyWakeupSong(@RequestBody request: ApplyWakeupSongRequest) =
        wakeupSongUseCase.applyWakeupSong(request)

    @UserAccess(roles = [RoleType.STUDENT])
    @PostMapping("/keyword")
    fun applyByKeyword(@RequestBody request: ApplyWakeupSongByKeywordRequest) =
        wakeupSongUseCase.applyByKeyword(request)

    @UserAccess(roles = [RoleType.TEACHER])
    @PatchMapping("/allow/{id}")
    fun allow(@PathVariable id: Long) = wakeupSongUseCase.allow(id)

    @UserAccess(roles = [RoleType.TEACHER])
    @PatchMapping("/deny/{id}")
    fun deny(@PathVariable id: Long) = wakeupSongUseCase.deny(id)

    @UserAccess(hasAnyRoleOnly = true)
    @DeleteMapping("/my/{id}")
    fun deleteMyWakeupSong(@PathVariable id: Long) = wakeupSongUseCase.deleteMyWakeupSong(id)
}
